mlir/lib/Transforms/Utils/LoopInvariantCodeMotionUtils.cpp

```diff
@@ -106,13 +175,18 @@ size_t mlir::moveLoopInvariantCode(
 size_t mlir::moveLoopInvariantCode(LoopLikeOpInterface loopLike) {
   return moveLoopInvariantCode(
       loopLike.getLoopRegions(),
-      [&](Value value, Region *) {
-        return loopLike.isDefinedOutsideOfLoop(value);
+      [&](Value value, Region *region) {
+        return !region->isAncestor(value.getParentRegion());
       },
       [&](Operation *op, Region *) {
-        return isMemoryEffectFree(op) && isSpeculatable(op);
+        return isSpeculatable(op) &&
+               (isMemoryEffectFree(op) || hasOnlyReadEffect(op));
```

Comment:
what if there are other operations aliasing the read-only op. That doesn't sound like a safe transformation?

Reply:
We thought about aliasing as well. For now, instead of a more complex analysis, we opt a more conservative approach that checks if there is any op with write and/or unknown side effects in the loop. If that's the case, then only pure ops get hoisted without any guard. 

This check is done [here](https://github.com/llvm/llvm-project/pull/120302/files#diff-a18ca7e24c6603fc6dbf7a1e8ea876c7cce7a9d44b7ed189720e1fd11467c645R143)